### ISSUE TITLE
Editor: Exiting block editor results in "oh hai" error on Jetpack site

### ISSUE THREAD

**Issue by rachelmcr** [Automattic/wp-desktop#941]
#### Steps to reproduce
1. Log in to the app with a WordPress.com account linked to a self-hosted Jetpack site.
2. Open the dev console (`CTRL + SHIFT + i` on macOS). This is not required to reproduce the issue but will allow you to see console errors when the issue occurs.
3. Go to My Sites and select your Jetpack site.
4. Click the "Write" button to start a new post in the block editor. (Alternately, open an existing post in the block editor.)
5. Click the "W" icon in the top left to exit the editor.

#### What I expected

I expected to exit the editor.

#### What happened instead

I get the "Oh hai" error screen and this error in the console:

```
VM533 desktop-app.js:78 Uncaught ReferenceError: logger is not defined
    at startDesktopApp (desktop-app.js:78)
    at desktop-app.js:176
```

This is also a really confusing screen because clicking the "Get in touch" button on that screen takes me to the [Automattic Work With Us](https://automattic.com/work-with-us/) page, but within the app (not in a web browser). The only way to get back to Calypso in the app is to click on the support bubble in the bottom right and select "More Help."

#### App Version / OS version

WPDesktop 6.0.1-beta2 / macOS 10.15.5

Using a self-hosted Jetpack site with SSO enabled, with the block editor enabled from the web.

#### Screenshot / Video

![wpdesktop-exitjpeditor](https://user-images.githubusercontent.com/8658164/87562978-9b4bb900-c6b6-11ea-8461-5893c0a9b46f.gif)

#### Context / Source
#beta-testing

**Comment by nsakaimbo** [comment]
Thanks for the report! Definitely something we should fix before this release gets out the door. The additional detail is super useful, and good to know we can repro reliably.

**Comment by rachelmcr** [comment]
I tested again in version 6.0.1-beta3 with a few test sites and got the same results. More details about the sites I tested with:

* WordPress.com Simple site (doesn't have this issue) - Gutenberg 8.3.0.
* WordPress.com Atomic site - WordPress 5.4.2, Jetpack 8.7.1 (updated July 15), using built-in Gutenberg (not sure the version).
* Self-hosted Jetpack site - WordPress 5.4.2, Jetpack 8.7.1 (updated July 14), using built-in Gutenberg (not sure the version).

On the self-hosted Jetpack site I tested with in the original report, I'm no longer able to load Gutenberg in the app. Every time I try, it times out with `REASON_BLOCK_EDITOR_UNKOWN_IFRAME_LOAD_FAILURE`. (Site has WordPress 5.4.2, Jetpack 8.7.1, and tested with built-in Gutenberg or Gutenberg 8.5.1.)

I noticed that Calypso changed the way this close button is added in https://github.com/Automattic/wp-calypso/pull/43313, so this is possibly related. However, that doesn't explain reports of other issues with the editor close button appearing now in previous versions of the desktop app (as raised internally: p1595333404365400-slack-apps).

**Comment by nsakaimbo** [comment]
Update: seems there's been more than one Jetpack regression driven by server-side logic: @kraftbj  has [a PR](https://github.com/Automattic/jetpack/pull/16544) up in the Jetpack repository to patch a runtime error that occurs in the Desktop app when Gutenberg is loaded.

**Comment by nsakaimbo** [comment]
> On the self-hosted Jetpack site I tested with in the original report, I'm no longer able to load Gutenberg in the app. Every time I try, it times out with REASON_BLOCK_EDITOR_UNKOWN_IFRAME_LOAD_FAILURE

I'm assuming SSO is enabled for the site on the web? Haven't encountered this as Gutenberg is still working for the J-Ninja sites I've set up. However it's becoming apparent there are a few Jetpack-related regressions at play at the moment but not sure if related to this specific issue.

**Comment by nsakaimbo** [comment]
oh, and just an FYI - looks like the `⌘ + 1` shortcut does work to exit the `hey.html` view to go back to the main landing page, so at least there's that escape hatch for debugging purposes.

**Comment by p-jackson** [comment]
I was testing to see whether this was related to the change I made in Automattic/wp-calypso#43313. I don't _think_ it is because I'm reproducing the issue even when the new overriding code isn't kicking in (it will only kick in when the version of Gutenberg is high enough, and a vanilla jurassic.ninja site doesn't install the GB plugin). I can also see in the element inspector that it's not using the new slot API.

I also think the navigation back to the posts page is working, and it's an error on posts page that causes the "Oh hai" screen to show. I put a break point at `desktop-app.js:131` and clicked the (W) button. When it breaks in to the debugger the `electron` global is undefined, which is why the "Oh hai" screen displays, it thinks it's being opened in a browser window.

While paused in the debugger, running `window.top === window` is false. So I'm thinking the navigation back to the posts page is happening, but it's happening in the iframe, not the parent frame.

**Comment by p-jackson** [comment]
New theory, it looks like I did in fact break this in Automattic/wp-calypso#43313.

I removed the old click override with with the new slot API: https://github.com/Automattic/wp-calypso/pull/43313/files#diff-9a28151776793191d27d22c072f05340L594
The idea for old clients being that the code that override's the `<a>`'s `href` would be enough. You can see that this code overrides the `href` and `target` to ensure the link opens in the parent frame 25c2b-pb.

However in this context it looks like the code that's running is here https://github.com/Automattic/jetpack/blob/15294dbfe8ee61ca649fe5891b9e4d0cc8702091/modules/calypsoify/mods-gutenberg.js#L33
And it's only overriding the `href`, _not_ the `target`. That means the "View Posts" page tries to load inside the iframe where `electron` isn't defined, and that causes the error.

Even though it's JavaScript, this is breaking old clients because `iframe-bridge-server.js` is served over the network from `widgets.wp.com` so that it Jetpack sites are up to date with Calypso.

I think ideally `mods-gutenberg.js` would override the `target` too, but that's not going to fix the issue until everyone updates their version of Jetpack. I think it's going to be safe to re-add the click overriding code to `iframe-bridge-server.js` because hopefully if the slot API is present it'll then just replace those changes entirely 🤞 

Creating PR to confirm.

**Comment by p-jackson** [comment]
Got a PR that I think fixes the issue Automattic/wp-calypso#44331

**Comment by p-jackson** [comment]
I'm seeing the fix appear in other environments, but not in wp-desktop. Not sure why that is yet.

**Comment by p-jackson** [comment]
The desktop app is still requesting this file: https://widgets.wp.com/wpcom-block-editor/calypso.editor.min.js?ver=20200722, specifically with the version `20200722`. The latest version of this file includes the fix.
I don't know what it takes to get this version number to update in the desktop app. I would have thought it was coming from wpcom, but maybe in this instance it comes from the Jetpack plugin?

**Comment by dereksmart** [comment]
> I don't know what it takes to get this version number to update

Yeah, this is [coming from Jetpack](https://github.com/Automattic/jetpack/blob/master/modules/wpcom-block-editor/class-jetpack-wpcom-block-editor.php#L283). The `version` is the `Ymd` date, so it should cycle after 24hrs and we'll be able to see if it resolves in the desktop app then.

**Comment by susanjanec** [comment]
I had a user report this issue in 3164223-zen. They provided a [video](https://monosnap.com/file/NCyRRzlXpokSjNNR1zu5lAS5mGk2fr) of this happening, although I wasn't able to reproduce on my end.

Had the user keep the Gutenberg plugin activated which fixed that issue.

Slack [thread](https://a8c.slack.com/archives/C3GP81E05/p1595429341361100) for reference.

**Comment by nsakaimbo** [comment]
Just did some additional testing of the fix in Automattic/wp-calypso/pull/44331 and confirmed it does address this bug in the desktop app. 🎉  However it will be another day or so before the updated file is propagated from the wpcom server to existing clients via Jetpack.
